### DISCUSSION TITLE
Front-end image compression before image component upload

### DISCUSSION THREAD

**Discussion by dashiell-zhang** [ant-design-blazor/ant-design-blazor#3416] [Q&A] [answered]
![image](https://github.com/ant-design-blazor/ant-design-blazor/assets/36178221/c53699e9-18a6-4dc9-a08a-29ba89bb39a0)
我目前有使用图片组件，目前遇到的问题是客户每次提交的图片都太大了，往往在 20Mb 左右每个图片，所以想要在 Blazor WASM 提交到后台 API 之前对图片进行一次压缩分辨率，修改图片分辨率的方法我已经实现了，目前遇到的问题是，无法将修改后的新图片信息覆盖到 file 中。

I currently use the image component, and the problem I am currently encountering is that the images submitted by customers are too large each time, usually around 20Mb per image, so I want to compress the image resolution before Blazor WASM submits it to the background API. I have already implemented the method of modifying the image resolution, but the problem I am currently encountering is that the modified new image information cannot be overwritten in the file.

![image](https://github.com/ant-design-blazor/ant-design-blazor/assets/36178221/7fb10dcf-f75a-4791-b34e-9e620ae5cc63)

**Comment by dashiell-zhang** [comment]
![屏幕截图 2023-07-04 231719](https://github.com/ant-design-blazor/ant-design-blazor/assets/36178221/ea00e97a-f2d8-45bd-8e87-d12076f14d2a)

我尝试了将新的 img 插入到 blob 中，将新的 blob url 复制给 file.ObjectURL 但是最后在组件提交的时候用的还是旧的 file.ObjectURL

I tried inserting the new img into the blob, copying the new blob url to file.ObjectURL but in the end the old file.ObjectURL was used when the component was submitted

**Comment by dashiell-zhang** [comment, marked as answer]
**Install SkiaSharp.Views.Blazor**

下面是我采用 input file 原生上传实现的方式，上传之后会检测图片的高和宽，将 高或宽的最大值裁剪为1920px，成功运行。

The following is the way I upload the input file natively. After uploading, the height and width of the image will be detected, and the maximum height or width will be cropped to 1920px. It runs successfully.

```
<FormItem Label="上传">
            <InputFile OnChange="@LoadFiles" multiple />
</FormItem>
```

  ```
  private async void LoadFiles(InputFileChangeEventArgs e)
    {
        foreach (var file in e.GetMultipleFiles())
        {
            string url = "https://xxxx.com/api/File/UploadFile";

            using HttpClientHandler handler = new();

            using HttpClient client = new(handler);
            client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher;

            foreach (var header in headers)
            {
                client.DefaultRequestHeaders.Add(header.Key, header.Value);
            }

            string boundary = "----" + DateTime.UtcNow.Ticks.ToString("x");

            using MultipartFormDataContent formDataContent = new(boundary);

            byte[] imageByteArray;

            using (var memoryStream = new MemoryStream())
            {
                await file.OpenReadStream(1024*10000).CopyToAsync(memoryStream);
                imageByteArray = memoryStream.ToArray();
            }

            using var original = SKBitmap.Decode(imageByteArray);

            int maxSideLength = 1920;

            int width = 0;
            int height = 0;

            if (original.Width > original.Height)
            {
                if (original.Width > maxSideLength)
                {
                    var percent = maxSideLength / (float)original.Width;
                    width = (int)(original.Width * percent);
                    height = (int)(original.Height * percent);
                }
            }
            else
            {
                if (original.Height > maxSideLength)
                {
                    var percent = maxSideLength / (float)original.Height;
                    width = (int)(original.Width * percent);
                    height = (int)(original.Height * percent);
                }
            }

            using var resizeBitmap = original.Resize(new SKImageInfo(width, height), SKFilterQuality.High);
            using var image = SKImage.FromBitmap(resizeBitmap);
            using var imageData = image.Encode(SKEncodedImageFormat.Jpeg, 100);

            var fileContent = new StreamContent(imageData.AsStream());

            fileContent.Headers.ContentType = new MediaTypeHeaderValue(file.ContentType);
            formDataContent.Add(fileContent, "file", file.Name);

            var httpResponse = await client.PostAsync(url, formDataContent);

            string responseStr = httpResponse.Content.ReadAsStringAsync().Result;

            Console.WriteLine("Success");
        }
    }
```

**Comment by Abin-Liu** [comment]
我会建议你在前端把压缩后的图片转换成base64字符串通过ajax传到服务器，服务端将base64复原成文件写入磁盘。

**Comment by dashiell-zhang** [comment]
早上好 @Abin-Liu ，个人感觉为了上传文件，把文件转成base64编码然后在服务器上转换回来，似乎这是一个没有意义的操作，目前常见的 对象存储服务，都不会这样做。
